packages/utilities/src/selection/Selection.ts

```diff
@@ -36,8 +36,8 @@ export class Selection implements ISelection {
   private _unselectableCount: number;
   private _isModal: boolean;
 
-  constructor(options: ISelectionOptions = {}) {
-    const { onSelectionChanged, getKey, canSelectItem = (item: IObjectWithKey) => true, selectionMode = SelectionMode.multiple } = options;
+  constructor(options: ISelectionOptions<TItem> = {}) {
```

Comment:
Will the compiler let us do:
```ts
constructor(options: TItem extends IObjectWithKey ? ISelectionOptions<TItem> : ISelectionOptionsWithOptionalGetKey<TItem> & ISelectionOptionsWithRequiredGetKey<TItem>)
```
This will ensure that if the supplied type doesn't have `key`, then the caller must supply the implementation.

Reply:
I'm not sure how to express that `options` is optional unless `TItem` doesn't extend `IObjectWithKey`.

Reply:
Alternatively, to continue to enforce the case even when `options` is not passed, do:
```ts
constructor(
  // If the item type has a built-in key...
  ...args: TItem extends IObjectWithKey ?
    // Then the arguments can be empty or have the options without `getKey`
    ([ISelectionOptions<TItem>] | []) :
    // Otherwise, arguments require options with `getKey`.
    [ISelectionOptionsWithRequiredGetKey<TItem>]
) {
  const [options] = args;
  ...
}
```

Reply:
Updated the comments a bit more. Let me know what you think--it's fairly verbose in the interest of being clear to consumers who may not be as good with reading TS types.